### ISSUE TITLE
PROXY protocol support

### ISSUE THREAD

**Issue by cperciva** [Tarsnap/spiped#134]
At some point, we may wish to add PROXY protocol support to spiped.  I don't think there's any urgency to this, but I'm opening an issue so I don't forget about it.

**Comment by TimWolla** [comment]
In fact I already thought about suggesting / implementing this in the past. Here's some comments on how this feature should work in my opinion:

Sending the PROXY header should be supported for both an encrypting and decrypting spiped:

For the decrypting end you might want to know which the client is that connects using the secret (probably especially helpful in combination with spipe).
For the encrypting end you get the classic PROXY protocol support that sends the IP address of the client (probably less important, as you would not expose an encrypting spiped to the Internet in most cases).

IMO there is no need to add a parser for the proxy protocol as spiped does not log any information anyway. If a service in front of the encrypting spiped wants to use the proxy protocol those bytes are simply forwarded and thus correctly interpreted by the service behind the decrypting spiped.

TL;DR: Add a single `-P` option that sends a PROXY header into the outgoing pipe, regardless whether you're encrypting or decrypting should be enough.

**Comment by cperciva** [comment]
Agreed.  The only slightly tricky bit is that the PROXY header needs to go into the plaintext stream -- when decrypting this means simply sending the data, but when encrypting this means encrypting it into a block and sending that.  So the PROXY header data would need to be injected at two different points. :-(

**Comment by TimWolla** [comment]
Oh and what just comes to my mind, you might need some interpretation of proxy protocol, because the [specification](http://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) states that:

> The sender must ensure that all the protocol header is sent at once.

Or does spiped already ensure that a single encrypted packet results in a single decrypted packet on the receiving side? Otherwise the PROXY header of an encrypting spiped might be split up by an decrypting spiped.

**Comment by cperciva** [comment]
It will be passed via a single syscall.  What happens in the TCP stack after that point we have no control over; but nobody else using the PROXY protocol has control either, so we're no worse off than anyone else.

**Comment by testcaoy7** [comment]
Using 3proxy at server end as spiped target can easily do this.

**Comment by Vesnica** [comment]
I use script below to enable connect through http proxy tunnel:
```
#!/bin/bash

#sudo apt-get install spiped libnet-proxy-perl
#dd if=/dev/urandom bs=256 count=1 of=keyfile

PIDFILE=spiped.pid

killall -w connect-tunnel
nohup connect-tunnel -U "Mozilla/4.03 [en] (X11; I; Linux 2.1.89 i586)" -P PROXY_IP:PROXY_PORT -T 23128:REMOTE_IP:REMOTE_PORT &

if [ -f $PIDFILE ]; then
    pkill -F $PIDFILE
    rm -f $PIDFILE
fi
spiped -e -s '[0.0.0.0]:35557' -t '[127.0.0.1]:23128' -k keyfile -p $PIDFILE
```
But this method is rather awkward, and connect-tunnel's performance is terrible.

It'd be great if I can simply wrote:
```
spiped -e -s '[0.0.0.0]:35557' -t '[127.0.0.1]:23128' -k keyfile -tunnel PROXY_IP:PROXY_PORT
```

**Comment by testcaoy7** [comment]
However, it will be amazing if spiped can have native proxy support.

**Comment by topmask** [comment]
> 

 I need to send the SPIPED encryption on the client, then SPIPED decryption to the 3proxy proxy to the website, then return the data to SPIPED encryption, and then SPIPED client decryption. How to deal with it?

**Comment by cperciva** [comment]
> I need to send the SPIPED encryption on the client, then SPIPED decryption to the 3proxy proxy to the website, then return the data to SPIPED encryption, and then SPIPED client decryption. How to deal with it?

I have no idea what you're asking here, or what it has to do with this github issue.
